### PR TITLE
Apply RLS to @splat builtin, eliminating its length parameter

### LINKED ISSUE #16277: Proposal: apply RLS to @splat, eliminating its length parameter
Status quo:
```zig
const x = @splat(4, @as(u32, 123)); // .{ 123, 123, 123, 123 }
```
Proposal:
```zig
// result type used to infer length and elem type
const x: @Vector(4, u32) = @splat(123); // .{ 123, 123, 123, 123 }
```
This brings similar advantages to #5909, and fits nicely with it. `@splat`'s current usage is quite unpleasant for constants in particular, due to the need to annotate the element type with `@as` - this would reduce visual noise in most cases.

### PR DESCRIPTION
**Pull request by antlilja**
Closes #16277

This PR removes the length parameter from the splat builtin instead resolving the result type.

### CODE REVIEW

File: src/AstGen.zig [hunk 1]
```diff
@@ -8591,10 +8591,10 @@ fn builtinCall(
         },
 
         .splat => {
-            const len = try expr(gz, scope, .{ .rl = .{ .coerced_ty = .u32_type } }, params[0]);
-            const scalar = try expr(gz, scope, .{ .rl = .none }, params[1]);
+            const scalar = try expr(gz, scope, .{ .rl = .none }, params[0]);
```
Comment (andrewrk) [inline src/AstGen.zig:8594]:
```suggestion
            const scalar = try comptimeExpr(gz, scope, .{ .rl = .none }, params[0]);
```

Would you mind tacking on this change and adding a behavior test for it?

Behavior test would look something like this:

```zig
test {
    @splat(foo());
}
```

It should make foo() be called at comptime even though the comptime keyword is not used. Same as it already does for e.g. `[foo()]T` (array type syntax).

**Comment by mlugg** [comment]
Sema logic is missing type checks. Need to check dest ty is a vector

**Comment by mlugg** [comment]
Also, CI failures are because zig1 doesn't have the changes. Run `zig build update-zig1` and commit the changes in the `stage1` dir (Andrew will recreate this commit before merging for security reasons)

File: src/AstGen.zig (hunk 1 above)
Comment (mlugg) [reply src/AstGen.zig:8594]: I'm confused, why would you do this? It's perfectly valid for the value being splat-ed to be runtime known; this isn't a length.

Comment (andrewrk) [reply src/AstGen.zig:8594]: My mistake, for some reason I thought it was a length, but that's very wrong. I apologize; ignore me!

**Comment by andrewrk** [comment]
Thanks for implementing this.

Hmmmm, looking at the diff, is this change actually good? It seems kind of worse actually. Are there any more opportunities to clean up the diff so that we can see if this change actually results in more maintainable code?

**Comment by mlugg** [comment]
I think a key thing that would make this nicer to work with is if we can figure out RLS for arithmetic operands, but I'm still unsure what exactly such a proposal would look like

**Comment by antlilja** [comment]
> Thanks for implementing this.
> 
> Hmmmm, looking at the diff, is this change actually good? It seems kind of worse actually. Are there any more opportunities to clean up the diff so that we can see if this change actually results in more maintainable code?

I've now pushed some new changes where I cleaned up the code as much as possible. 

I do agree that some parts do look a bit less readable. This is the resulting change when the result type cannot be inferred:

```zig
// Before:
@splat(N, @as(T, SOME_LITERAL));

// After:
@as(@Vector(N, T), @splat(SOME_LITERAL));
```

In quite a few places the code is less readable because the type cannot be inferred where it probably could be. I think @mlugg is correct in that RLS for arithmetic operands would help but there are a lot of places where this wouldn't help. For example where this syntax change interacts with the new cast syntax:

```zig
// Before:
@splat(len, @as(i32, @intCast(some_var)))

// After:
@as(@Vector(len, i32), @splat(@as(i32, @intCast(some_var))))
```

**Comment by mlugg** [comment]
That case can be fixed by adding a result type for `@splat`'s operand using a `elem_type_index` instruction (or, similarly, by by extending `elem_type` to work with vectors).

**Comment by andrewrk** [comment]
Thanks for further exploring it with cleanups, @antlilja. Upon reexamination, I think that this change is at worst, neutral. Given that it increases consistency with other builtins, and sometimes can reduce an awkward need to pass the element type rather than the vector type, I'm in favor of moving forward with it.

**Comment by antlilja** [comment]
> Thanks for further exploring it with cleanups, @antlilja. Upon reexamination, I think that this change is at worst, neutral. Given that it increases consistency with other builtins, and sometimes can reduce an awkward need to pass the element type rather than the vector type, I'm in favor of moving forward with it.

Then I'll keep working on it. I've started implementing what mlugg suggested so it works better with the other changes to builtins.

**Pull request merged**